clang/docs/checkedc/Bounds-Checking-for-LValue-Expressions.md

```diff
@@ -0,0 +1,268 @@
+# Bounds Checking: Generalizing from Variables to LValues
+
+## Problem Summary
+
+The current bounds checking algorithm checks that the inferred bounds of each
+in-scope variable imply the target bounds of the variable after each top-level
+CFG statement. This supports bounds checking in multiple assignments.
+For example:
+
+```
+void f(array_ptr<int> small : count(1),
+       array_ptr<int> medium : count(2),
+       array_ptr<int> large : count(3)) {
+	// At the end of medium = small: medium's inferred bounds are count(1).
+	// These inferred bounds do not imply the target bounds of count(2).
+	// At the end of medium = large: medium's inferred bounds are count(3).
+	// These inferred bounds imply the target bounds of count(2).
+	// The bounds are not checked until the end of the statement, when
+	// the inferred bounds of count(3) imply the target bounds of count(2).
+	medium = small, medium = large;
+}
+```
+
+For other kinds of lvalue expressions (`*p`, `a.f`, etc.), bounds checking
+occurs immediately after checking an assignment to the expression. The goal
+is to generalize the work that was previously done for variables so that the
+bounds checking behavior is the same for all lvalue expressions.
+
```

Comment:
These are checked in `TraverseTopLevelVarDecl`. This method itself doesn't do anything to the `CheckingState` members, including `ObservedBounds` and `EquivExprs` - it just initializes an empty `CheckingState` and checks the bounds immediately after traversing its `VarDecl *` argument.

How much detail on bounds checking do you think would be helpful in this document? Initially I didn't include `TraverseTopLevelVarDecl` here since it doesn't directly interact with `ObservedBounds` (or `EquivExprs`). I have another document in progress in #945 that describes bounds checking in general. I would propose that I move most of the detail in this "Bounds Checking for Variables" section into that document so that this document can focus more on lvalue generalization. The "Bounds Checking for Variables" section in this document can then just briefly list the places where `ObservedBounds` is updated or used since this is relevant to changing the `ObservedBounds` keys, and reference the other bounds checking document for more detail. What are your thoughts on this?

Reply:
Yes, I agree that most of the detail on bounds checking can be in the other document (#945) and this document can focus on lvalue generalization.

Reply:
Thanks. I've moved most of this section into the #945 document and just included a list of methods where it will be necessary to change the keys of `ObservedBounds` in this section